File: sdks/php/src/CalendarEvent.php
<?php

declare(strict_types=1);

namespace Mobiscroll\Connect;

/**
 * @property-read array<string, mixed> $attendees
 * @property-read array<string, mixed> $custom
 */
class CalendarEvent
{
    /**
     * @param array<string, mixed> $original
     * @param array<string, mixed>|null $attendees
     * @param array<string, mixed>|null $custom
     */
    public function __construct(
        public readonly string $provider,
        public readonly string $id,
        public readonly string $calendarId,
        public readonly string $title,
        public readonly \DateTime $start,
        public readonly \DateTime $end,
        public readonly bool $allDay,
        public readonly ?string $recurringEventId = null,
        public readonly ?string $color = null,
        public readonly ?string $location = null,
        public readonly ?array $attendees = null,
        public readonly ?array $custom = null,
        public readonly ?string $conference = null,
        public readonly ?string $availability = null,
        public readonly ?string $privacy = null,
        public readonly ?string $status = null,
        public readonly ?string $link = null,
        public readonly array $original = [],
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        return new self(
            $data['provider'] ?? throw new \InvalidArgumentException('provider is required'),
            $data['id'] ?? throw new \InvalidArgumentException('id is required'),
            $data['calendarId'] ?? throw new \InvalidArgumentException('calendarId is required'),
            $data['title'] ?? '',
            new \DateTime($data['start'] ?? 'now'),
            new \DateTime($data['end'] ?? 'now'),
            $data['allDay'] ?? false,
            $data['recurringEventId'] ?? null,
            $data['color'] ?? null,
            $data['location'] ?? null,
            $data['attendees'] ?? null,
            $data['custom'] ?? null,
            $data['conference'] ?? null,
            $data['availability'] ?? null,
            $data['privacy'] ?? null,
            $data['status'] ?? null,
            $data['link'] ?? null,
            $data['original'] ?? [],
        );
    }
}

File: sdks/dotnet/src/Mobiscroll.Connect/Models/CalendarEvent.cs
using System;
using System.Collections.Generic;
using System.Text.Json;

namespace Mobiscroll.Connect.Models;

public sealed class CalendarEvent
{
    public Provider Provider { get; set; }
    public string Id { get; set; } = string.Empty;
    public string CalendarId { get; set; } = string.Empty;
    public string Title { get; set; } = string.Empty;

    /// <summary>Event description or notes.</summary>
    public string? Description { get; set; }

    public DateTime Start { get; set; }
    public DateTime End { get; set; }
    public bool AllDay { get; set; }
    public string? RecurringEventId { get; set; }
    public string? Color { get; set; }
    public string? Location { get; set; }
    public List<EventAttendee>? Attendees { get; set; }
    public Dictionary<string, JsonElement>? Custom { get; set; }
    public string? Conference { get; set; }

    /// <summary>
    /// Provider-specific conference metadata; use it for details beyond <see cref="Conference"/>.
    /// Keys vary per provider, so the payload is left as raw <see cref="JsonElement"/> values.
    /// </summary>
    public Dictionary<string, JsonElement>? ConferenceData { get; set; }

    /// <summary>"busy" | "free"</summary>
    public string? Availability { get; set; }

    /// <summary>"public" | "private" | "confidential"</summary>
    public string? Privacy { get; set; }

    /// <summary>"confirmed" | "tentative" | "cancelled"</summary>
    public string? Status { get; set; }

    /// <summary>ISO 8601 timestamp of the last modification, e.g. "2026-03-10T13:36:08.000Z".</summary>
    public string? LastModified { get; set; }

    public string? Link { get; set; }

    /// <summary>Raw provider event payload.</summary>
    public JsonElement? Original { get; set; }
}
